### PR TITLE
Implement a Kraken price source for tokens.

### PR DESCRIPTION
**Pull request by nlordell**
This PR uses the Kraken API implemented in #562 in order to retrieve prices from a list of token infos.

### Test Plan

New unit tests. Also, a test that does not run by default is included and can be run by:
```
cargo test online_kraken_prices -- --ignored --nocapture
```

### CODE REVIEW

File: driver/src/price_estimation/mod.rs [hunk 1]
```diff
@@ -1,4 +1,71 @@
...
+        }
+    }
+}
+
+/// An abstraction around a type that retrieves price estimate from a source
+/// such as an exchange.
+#[cfg_attr(test, mockall::automock)]
+trait PriceSource {
+    /// Retrieve current prices for the current tokens. Returns a sparce price
+    /// array as being unable to find a price is not considered an error.
```
Comment (e00E) [inline driver/src/price_estimation/mod.rs]: Mention in what unit the price is measured.

**Review by fleupold — APPROVED** [review]
Spot on! Very nice...

<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia2.giphy.com%2Fmedia%2FKAe3Ez73EnmbDh9rot%2Fgiphy.gif&token=bullseye" data-gitmeme-token="bullseye"></a>

File: driver/src/price_estimation/kraken.rs [hunk 2]
```diff
@@ -1,4 +1,203 @@
...
+    api: Api,
+}
+
+/// The default Kraken API base URL.
+const DEFAULT_API_BASE_URL: &str = "https://api.kraken.com/0/public";
+
+impl KrakenClient<KrakenHttpApi> {
+    /// Creates a new client instance using an HTTP API instance and the default
+    /// Kraken API base URL.
+    pub fn new() -> Result<Self> {
```
[the remark this answers is not in the document]
Comment (nlordell) [reply driver/src/price_estimation/kraken.rs]: `Default` must return `Self` and doesn't support `Result<Self>`. Unfortunately, there is no `TryDefault` trait in `std`.

File: driver/src/price_estimation/kraken.rs [hunk 3]
```diff
@@ -1,4 +1,203 @@
...
+
+/// A client to the Kraken exchange.
+pub struct KrakenClient<Api> {
+    /// A Kraken API implementation. This allows for mocked Kraken APIs to be
+    /// used for testing.
+    api: Api,
+}
+
+/// The default Kraken API base URL.
+const DEFAULT_API_BASE_URL: &str = "https://api.kraken.com/0/public";
```
[the remark this answers is not in the document]
Comment (nlordell) [reply driver/src/price_estimation/kraken.rs]: :eyes:  Oops.

File: driver/src/price_estimation/kraken.rs [hunk 4]
```diff
@@ -1,4 +1,200 @@
...
+        KrakenClient { api }
+    }
+
+    /// Generates a mapping between Kraken asset pair identifiers and tokens
+    /// that are used when computing the price map.
+    fn get_token_asset_pairs<'a>(&self, tokens: &'a [Token]) -> Result<HashMap<String, &'a Token>> {
+        // TODO(nlordell): If these calls start taking too long, we can consider
+        //   caching this information somehow. The only thing that is
+        //   complicated is determining when the cache needs to be invalidated
+        //   as new assets get added to Kraken.
```
Comment (bh2smith) [inline driver/src/price_estimation/kraken.rs]: Is there an open issue for this TODO?

Comment (nlordell) [reply driver/src/price_estimation/kraken.rs]: There was, but there was also a typo in the title: #583

Comment (nlordell) [reply driver/src/price_estimation/kraken.rs]: Not for this particular `TODO`. Maybe a `NOTE` would be better. I just left it there as a possible optimization if delays become an issue in the future.

There is a related case here: #583

**Pull request merged**